torch/_decomp/decompositions.py

```diff
@@ -1086,7 +1086,7 @@ def dropout(input: Tensor, p: float, train: Optional[bool]):
     if train and p != 0:
         return aten.native_dropout(input, p, train)[0]
     else:
-        return input.clone()
+        return input
```

Comment:
This doesn't match the alias information for `aten::dropout` which claims to always return a new Tensor. 
```yaml
- func: dropout(Tensor input, float p, bool train) -> Tensor
```
However, I do see that in practice `dropout` does return an alias in eval mode. I think the correct fix here would be to move the fast-return into `dropount` definition in `torch/nn/functional.py` and leave the clone here.

Reply:
If we follow what you said, then `native_dropout` also needs to return a new Tensor.

```
- func: native_dropout(Tensor input, float p, bool? train) -> (Tensor, Tensor)
```

However, the current decomposition will return the input directly too.

```python
@register_decomposition(aten.native_dropout)
@out_wrapper("out0", "out1")
def native_dropout(input: Tensor, p: float, train: Optional[bool]):
    if train and p != 0:
        if p == 1:
            return (torch.zeros_like(input), torch.zeros_like(input, dtype=torch.bool))
        if not input.dtype.is_floating_point:
            raise RuntimeError(
                "result type Float can't be cast to the desired output type Long"
            )
        bool_mask = torch.rand_like(input) > p
        res = bool_mask * input * float(1.0 / (1.0 - p))
        return (res, bool_mask)
    else:
        return (input, torch.ones_like(input, dtype=torch.bool)) # <--- this line
```

Reply:
Well that decomposition is wrong as `native_dropout` actually does the clone
https://github.com/pytorch/pytorch/blob/c80e2d5bb23b48410dd73bc0698df8552f3dfaba/aten/src/ATen/native/Dropout.cpp#L123

Reply:
Emmmm we have so much inconsistency here.

Reply:
Added a bit more context here: https://github.com/pytorch/pytorch/issues/115249#issuecomment-1856319556 (but agreed with @peterbell10 )

Reply:
IMO the real source of truth is eager mode, not the alias information metadata.
```
x=torch.randn(100)
>>> x is torch.nn.functional.dropout(x, training=False, p=0.5)
True
>>> x is torch.nn.functional.dropout(x, training=True, p=0.0)
True
>>> x is torch.nn.functional.dropout(x, training=True, p=0.5)
False
>>> 
```
So the alias information pointed out by @peterbell10 is incorrect and should be changed.

Adding a clone in `torch.compile` mode, when eager mode doesn't seems silly.  First, it will mean we are slower than eager mode for no reason.  Second, if you return the output the aliasing of eager won't match the aliasing of compile, which could lead to bugs.

So as I see it there are basically two options that maintain consistency between eager/compile:
1) Remove the clone from the decomp, and update the signature of dropout to reflect that it aliases.  This will likely be a minor perf improvement in some cases.
2) Add a new clone to eager mode so the existing aliasing metadata is correct.  This will sightly regress eager mode inference performance.

I would lean towards option 1, which is this PR.  Though the remaining tests need to be fixed, and I would like the bypass of the duplicate pattern exclusion to be more narrow.  Simply turning off the check will lead to bugs.

Reply:
There is a third option which I suggested above. We move the aliasing fast-return to `F.dropout` and update the behavior of `aten.dropout` to never alias. This would maintain the current behavior in the python API, remove the unnecessary clone in the aot graph and make the historic aliasing info now correct for the ATen API.

The only practical difference between this and option 1 is whether we need to change the alias info for `aten.dropout`, which I was trying to avoid as I assumed it would be a BC break. Perhaps that's okay here since the alias info was wrong to begin with, though.

I don't think we should go with option 2 as this alaising behavior is very long-standing (in fact it looks like it predates aliasing info).

Reply:
What would making the signature accurate break?

Reply:
To resolve the dispute, maybe we should reach agreement on this question: https://dev-discuss.pytorch.org/t/what-is-the-relationship-among-nn-module-api-nn-functional-api-aten-api/1735 👀

Reply:
Oh sorry! I misread.

`aten.dropout` is a `CompositeImplicitAutograd` operator, and in general we don't (and sometimes cannot) guarantee proper aliasing info in their schemas. `aten.contiguous` is another example: it doesn't have accurate aliasing info in its schemas, because decides whether or not to create a fresh tensor or return the input directly based on the striding information of the input.

So.. I think having a decomp for `aten.dropout` return the input directly in one path is probably ok. It wouldn't be ok to do this, though, for `aten.native_dropout`, which is not a composite op, so it must have correct aliasing info.

One question that I have though is maybe for @andrewor14: I think one downsides of the current approach is that for someone performing export, if you exported an inference graph you will have no way of knowing that your model originally had a dropout in it (since it effectively no-ops in the inference case).

Reply:
Contiguous is registered with the correct aliasing info:
```yaml
- func: contiguous(Tensor(a) self, *, MemoryFormat memory_format=contiguous_format) -> Tensor(a)
```
where according the README,  `(a)` means _may_ alias, so this is accurate. I do agree that this doesn't matter nearly as much for `CompositeImplicitAutograd` operators though.

> What would making the signature accurate break?

If we add the same aliasing info to dropout, we necessarily change the operator schema. I would be concerned that any code that looks up the operator by its schema will break. e.g. TorchScript does this:

https://github.com/pytorch/pytorch/blob/bf20b56e9d40dd005399fce05b3a073d8132e04f/torch/csrc/jit/tensorexpr/lowerings.cpp#L27-L29

I would also be concerned about breaking backward compatibility for exported TorchScript models. I haven't tried this though, so maybe they aren't sensitive to the aliasing info in the schema.

Reply:
Ah you're right, `contiguous` is technically correct. `aten.batch_norm` is technically wrong though, since it may mutate the running stat inputs if the training flag is set (it decomposes into ops with correct schemas, at least): https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L1058

Reply:
Hrm and `aten.native_batch_norm` is also incorrect, whilst not even being `CompositeImplicitAutograd`. This is a bit of a mess :/

I also see it's been a year since this "In 2 weeks" comment:
https://github.com/pytorch/pytorch/blob/054f9548b481ecafd754c0780936e2c95ce1237a/aten/src/ATen/native/native_functions.yaml#L4238

@janeyx99 do you remember what happened to this?

Reply:
Yea, because the schema for native_batch_norm was incorrect and we didn't want to break BC, we introduced a native_batch_norm_legit that native_batch_norm decomposes into, which would have the right schema when torch.compile sees it (for correct functionalization).

The 2 weeks were due to the forward compat window internally, but I never ended up getting bandwidth to refactor completely :/ @andrewor14 is currently working on a refactor though, that should rectify and consolidate batch_norm once and for all!

Reply:
> One question that I have though is maybe for @andrewor14: I think one downsides of the current approach is that for someone performing export, if you exported an inference graph you will have no way of knowing that your model originally had a dropout in it (since it effectively no-ops in the inference case).

Yes, this is a problem for export + training use cases. We can't always rely on the user exporting in train mode initially. Also, even if they exported in train mode we still have a problem: switching between train/eval modes would require us to hardcode the aten op to replace, whereas today we express the match and replacement patterns as torch ops agnostic to how they will eventually be decomposed (see [here](https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/eval_utils.py#L26)).